Provided a better solution for reading dead outbox entries (#1940)
cap-js/docs: java/outbox.md

:::

### Reading Dead Entries

This filtering of dead entries is done on the database by adding `where` clauses for each outbox and their maximum number of retries. The following code provides the handler for the `DeadLetterQueueService` that modifies the where clause by adding
additional conditions for filtering the outbox entries:

```java
@Component
@ServiceName(OutboxDeadLetterQueueService_.CDS_NAME)
public class DeadOutboxMessagesHandler implements EventHandler {

	private final PersistenceService db;

    public DeadOutboxMessagesHandler(@Qualifier(PersistenceService.DEFAULT_NAME) PersistenceService db) {
        this.db = db;
    }

    @Before(entity = DeadOutboxMessages_.CDS_NAME)

>     public void modifyWhereClause(CdsReadEventContext context) {

Reviewer: ```suggestion
    public void addDeadEntryFilter(CdsReadEventContext context) {
```

>         CqnSelect cqn = context.getCqn();

Reviewer: ```suggestion
```

        Optional<Predicate> outboxFilters = this.createOutboxFilters(context.getCdsRuntime());
        CqnSelect modifiedCqn = copy(

>           cqn,

Reviewer: ```suggestion
          context.getCqn(),
```

          new Modifier() {
              @Override

>               public CqnPredicate where(Predicate where) {

Reviewer: much simpler logic:
- do the copy only in case of outboxFilters.isPresent()
- just return outboxFilters.get().and(where) in the modifier. Anding null should be fine.

                  if (where != null && outboxFilters.isPresent()) {
                      return where.and(outboxFilters.get());
                  } else if (where == null && outboxFilters.isPresent()) {
                      return outboxFilters.get();
                  } else if (where != null && !outboxFilters.isPresent()) {
                      return where;
                  } else {
                      return null;
                  }
              }
          });

        context.setCqn(modifiedCqn);
    }

    private Optional<Predicate> createOutboxFilters(CdsRuntime runtime) {

>         List<OutboxService> outboxServices = runtime.getServiceCatalog().getServices(OutboxService.class)

Reviewer: what about iterating the persistent outboxes from application configuration (outboxConfigs)+ the known defaults? INMEMORY_NAME would not be required.

Author: @BraunMatthias If an outbox has the default configuration, iterating over the configurations doesn't return the specific config for an outbox. I observed this while testing the code; in this case no outbox can be found and hence no dead entries are read at all.

                .filter(s -> !s.getName().equals(OutboxService.INMEMORY_NAME)).toList();
        CdsProperties.Outbox outboxConfigs = runtime.getEnvironment().getCdsProperties().getOutbox();

        Predicate where = null;
        for(OutboxService service : outboxServices) {

>             OutboxServiceConfig config = outboxConfigs.getService(service.getName());

Reviewer: why is config required at all?

Author: @BraunMatthias To get the configuration for the `maxAttempts`.

>             Predicate targetPredicate = CQL.get(Messages.TARGET).eq(service.getName()).and(CQL.get(Messages.ATTEMPTS).ge(config.getMaxAttempts()));

Reviewer: why do we need to filter by individual outbox and then OR the condition meeting all outboxes?

Author: @BraunMatthias Every outbox can have its own configuration for `maxAttempts`. To read only the red entries per outbox, this filter is required.

Reviewer: same reason - I just overlooked this aspect

[Learn more about CQL statement inspection.](./working-with-cql/query-introspection#cqnanalyzer){.learn-more}

::: tip Use paging logic

> Avoid to read all entries of the `cds.outbox.Messages` or `OutboxDeadLetterQueueService.DeadOutboxMessages` table at once, as the size of an entry is unpredictable and depends on the size of the payload. Prefer paging logic instead.

Reviewer: ```suggestion
Avoid to read all outbox entries at once as a single entry can have significant size reflecting the request's payload. Prefer `READ`-queries with paging instead.
```
